README.md
## Fuego on iOS

The upstream [Fuego](http://fuego.sourceforge.net/) project is a collection of C++ libraries for developing software for the game of Go.

The primary goal of the Fuego on iOS project is to provide a build of the Fuego source code for the iOS platform. The build is packaged into a framework bundle that can be easily integrated into any Xcode project. The build tries to remain as close as possible to the original upstream source code, with only such modifications as are necessary to overcome any difficulties in 1) building for the iOS platform, and in 2) building with a modern version of Boost.

A secondary goal of the Fuego on iOS project is to provide an iOS build of Fuego that is suitable for integration into the [Little Go](https://github.com/herzbube/littlego) project.

Fuego on iOS currently combines

* Boost 1.56.0
* Fuego trunk r1728

## Quickstart Guide

The project is set up with the `fuego-on-ios` branch as the default branch. This should allow you to get going immediately:

1. Clone the repository
1. Initialize the "modular-boost" submodule. Find the necessary commands further down in the section "[The modular-boost submodule](#the-modular-boost-submodule)".
1. Build Boost, then build Fuego. Find the necessary commands further down in the section "[How to build](#how-to-build)".

That's it, now all you need to do is integrate the Boost and Fuego frameworks into your project.

## Branches

##### Overview

The Fuego on iOS project has 3 permanent task branches:

* master
* fuego-on-ios
* fuego-for-littlego

The next sections explain the purpose of each of these task branches. There are also a number of branches that correspond to upstream Subversion branches. These branches are of no particular interest to the Fuego on iOS project and exist simply to mirror the upstream repository structure.

##### master

Fuego on iOS uses `svn2git` to mirror the [upstream Subversion repository](http://svn.code.sf.net/p/fuego/code/). The purpose of the `master` branch is to track the upstream `trunk`. No commits are allowed in `master` except those made by `svn2git` to synchronize with upstream. The reason for this hard rule is that `svn2git` will rebase local commits on top of Subversion commits, which would lead to problems due to published history being changed.

Currently, upstream changes are manually synchronized from time to time. A future goal is to automate this process (see issue #1 on GitHub).

##### fuego-on-ios

The `fuego-on-ios` branch is dedicated to maintaining the following stuff:

1. `boost/modular-boost`: A Git submodule that tracks a specific release of Boost in [this upstream repository](https://github.com/boostorg/boost).
1. `boost/boost.sh`: A script that builds Boost for the iOS platform and packages it into a framework bundle. The script is a derivate of Pete Goodliffe's "Boost on iPhone" script. The script was forked from [this popular repo](https://gitorious.org/boostoniphone/galbraithjosephs-boostoniphone) so that outstanding bugs could be fixed for the Fuego on iOS project.
1. `fuego-on-ios.xcodeproj`: The Xcode project that defines the build of Fuego for the iOS platform.
1. `build.sh`: A script that uses the Xcode project to build Fuego (via `xcodebuild`) and packages the result into a framework bundle.

Also in this branch are those changes to the original Fuego source code that are necessary to make Fuego compile in the Xcode environment.

##### fuego-for-littlego

The branch `fuego-for-littlego` contains those changes to the Fuego source code that are required by the [Little Go](https://github.com/herzbube/littlego) project.

These changes are separated into their own branch because the intent for the `fuego-on-ios` branch is to keep that build as close to the original source as possible so that other people can take it from there and add their own stuff without the Little Go modifications (which might not be palatable to everyone).

## Which branch should I use?

You should use `master` if you are a Git user who wants to work with the unmodified Fuego source code, but cannot be bothered to fiddle with Subversion or `svn2git`. The `master` branch simply takes care of "wrapping" the upstream Subversion repository in a Git repository. Currently the disadvantage is that the `master` branch may not be up-to-date because it is manually synchronized.

You should use `fuego-on-ios` if you want a ready-to-build Fuego + Boost environment. The build products are two framework bundles (Fuego + Boost) that you can simply add to your iOS application project.

You probably will not be interested in `fuego-for-littlego` at all, unless you want to study how the [Little Go](https://github.com/herzbube/littlego) app integrates Fuego.

## The modular-boost submodule

The `fuego-on-ios` branch (and also the `fuego-for-littlego` branch) contains a Git submodule in this project folder:

    boost/modular-boost

The submodule points to a specific Boost release in the official upstream Boost repository on GitHub (https://github.com/boostorg/boost). After you clone the Fuego on iOS repository, you must perform the following commands to also clone the submodule:

    cd /path/to/fuego-on-ios
    git submodule init
    git submodule update

**IMPORTANT:** Although the submodule tracks only a single tag of the upstream Git repository, cloning requires that the full upstream repo is replicated locally. Because the Boost project has a long history, the resulting download is quite large. At the time of writing the initial clone consumes roughly 480 MB of disk space. On a 3.0 mbps Internet connection this takes slightly more than 20 minutes to download.

## How to build

These are the commands to first build Boost, then build Fuego:

    cd boost
    ./boost.sh
    cd ..
    ./build.sh

These are the build results that can be integrated into other Xcode projects:

    boost/ios/framework/boost.framework
    ios/framework/fuego-on-ios.framework

Most important build settings:

* iOS SDK = The latest SDK known to your Xcode
* Deployment target = 7.0
* Architectures: armv7, armv7s, arm64 (iOS builds), i386, x86_64 (iPhone Simulator builds)
* C++ Language Dialect = GNU++98 (`-std=gnu++98`)
* C++ Standard Library = libstdc++ (`-stdlib=libstdc++`)
* Boost libraries: thread, filesystem, program_options, system, test, date_time

Environment variables that you can set and export to override build settings (both for the Boost and the Fuego build scripts):

* IPHONEOS_BASESDK_VERSION
* IPHONEOS_DEPLOYMENT_TARGET
* IPHONE_SIMULATOR_BASESDK_VERSION
* IPHONE_SIMULATOR_DEPLOYMENT_TARGET

## Repository maintenance

These are some notes on how to maintain the Fuego on iOS repository and its branches.

##### Synchronizing master with upstream Fuego

Checkout `master`, then run this command:

    svn2git --metadata --rebase

If you don't have `svn2git` in your environment, read the section "[Getting svn2git](#getting-svn2git)". If you *do* have svn2git, but running the command gives you an error message, you should probably read the section "[Reconnecting a cloned Git repository with upstream Subversion repository](#reconnecting-a-cloned-git-repository-with-upstream-subversion-repository)" further down.

In an ideal world, the `--rebase` command line option would update the local Git repository so that it is an exact mirror of the upstream Subversion repository. Unfortunately, in the `svn2git` version that is current at the time of writing (v2.2.2) the mirroring capability of `--rebase` is severely limited. In fact, the command line option does not do much else besides fetching upstream commits into the local Git repository. The following tasks need to be done manually:

1. Add local tags for tags that were created upstream since the last sync
    1. List all remote branches: `git branch -r`
    1. List all tags: `git tag`
    1. Pick out remote SVN branches whose name starts with `svn/tags` for which there is no corresponding tag
    1. For every branch thus found, get a log message: `git show svn/tags/[TAG-NAME]`
    1. Create the corresponding tag: `git tag -a -m "[LOG MESSAGE]" [TAG-NAME] svn/tags/[TAG-NAME]`
    1. Remove the remote branch: `git branch -rd svn/tags/[TAG-NAME]`
1. Add local branches for branches that were created upstream since the last sync
    1. List all branches: `git branch -a`
    1. Pick out remote SVN branches for which there is no corresponding local branch
    1. For every branch thus found, create a local branch: `git branch [BRANCH-NAME] remotes/svn/[BRANCH-NAME]`
1. Advance the HEAD of all existing branches that received commits upstream since the last sync. Usually, this affects the `master` branch.
   1. This affects all local branches where HEAD does not point to the same commit as HEAD of the corresponding remote branch
   1. Check out each affected branch, then advance its HEAD: `git reset --hard [COMMIT-REF]`

Once the local Git repository is in shape, the changes can be pushed to GitHub.


##### Integrating changes from master into fuego-on-ios

All commits in `master` up to the desired commit must be merged into the `fuego-on-ios` branch. To keep things simple no cherry-picking is allowed.

After the merge, the Xcode project may also need to be updated (add new files, remove old files, update build settings).

After `fuego-on-ios` has been updated, changes must be further merged into the `fuego-for-littlego` branch.

##### Upgrading to a new Boost release

To upgrade to a new Boost release, the `modular-boost` submodule must be changed so that it points to the commit of the Git tag that represents the desired release. This must be done on the `fuego-on-ios` branch. For instance, to upgrade to version 1.56.0:

    cd /path/to/fugo-on-ios
    git checkout fuego-on-ios
    cd boost/modular-boost
    git checkout boost-1.56.0
    cd ../..
    git add .
    git commit -m "upgrade submodule modular-boost to tag boost-1.56.0"

There is a possibility that the Fuego source code does not build with the new Boost release. If such a problem occurs, check with Fuego upstream if they already know a solution.

After `fuego-on-ios` has been updated, changes must be further merged into the `fuego-for-littlego` branch.

##### Making changes in fuego-for-littlego

All changes in the `fuego-for-littlego` branch remain local to that branch.

##### Updating this README

Updates to this README file are made on the `fuego-on-ios` branch and then merged into `fuego-for-littlego`. The README file does not exist in `master`.

## Troubleshooting

##### Getting svn2git

`svn2git` is a Ruby wrapper around `git svn`. Source code and installation instructions are available from [this GitHub repository](https://github.com/nirvdrum/svn2git). The Fuego on iOS project uses `svn2git` because it adds two bits of magic to `git svn`:

* `svn2git` creates native Git branches and tags from Subversion branches and tags, whereas `git svn` just imports everything into the `master` branch and creates remote tracking branches for Subversion branches and tags.
* `svn2git` points HEAD of `master` to the commit that represents the latest Subversion revision in the Subversion repo's trunk, whereas `git svn` points HEAD of `master` to the most recent Subversion revision, in whatever branch that may be (not necessarily trunk).

As of this writing, the latest released version of `svn2git` is 2.2.2. This version has a problem interacting with Git releases 1.8.3.2 and later. If you have this configuration, you can either downgrade Git on your system to 1.8.3.1, or you can get an unreleased version of svn2git where [this fix](https://github.com/nirvdrum/svn2git/issues/132) has been integrated.
  
##### Reconnecting a cloned Git repository with upstream Subversion repository

When you clone the Fuego on iOS repository from GitHub, you will not be able to use `svn2git` to synchronize the local repository with upstream Fuego. The reason is that certain configuration information required by `svn2git` (actually `git svn`) is kept only locally and is not versioned in Git, i.e. this information cannot be shared via GitHub. Therefore, anyone who wants to use `svn2git` must first reconnect their local clone of `fuego-on-ios` to the upstream Subversion repository.

The indicator that you are affected by this is when you issue a `svn2git` command and you get the error message "command failed: 2>&1 git svn fetch".

To fix the problem, add some configuration information to the `.git/config` file of your local repository:

    cd /path/to/fuego-on-ios
    git config svn-remote.svn.url http://svn.code.sf.net/p/fuego/code
    git config svn-remote.svn.fetch trunk:refs/remotes/trunk
    git config svn-remote.svn.branches branches/*:refs/remotes/*
    git config svn-remote.svn.tags tags/*:refs/remotes/tags/*

Now checkout `master` if you have not yet done so, then run

    svn2git --rebase

This takes quite a long time because it fetches all revisions from upstream Subversion. Behind the scenes the contents of the `.git/svn` folder are restored. From now on, `svn2git --rebase` should work as expected.

## Recipes

##### Build Fuego for Mac OS X

Assuming you have Boost installed somewhere locally (e.g. through Fink, MacPorts or Brew), you can run these commands to build and run Fuego for Mac OS X:

    cd /path/to/fuego-on-ios
    autoreconf -i
    ./configure --prefix $(pwd)/macosx
    make
    make install
    ./macosx/bin/fuego

Such a build may be useful to test patches on the command line.

##### How the Fuego on iOS repository was initialized

This purely historical information documents how the `fuego-on-ios` repository was initialized.

    mkdir fuego-on-ios
    cd fuego-on-ios
    svn2git --metadata http://svn.code.sf.net/p/fuego/code/

##### How the Xcode project was created

Project creation:

* Create new Xcode project with Xcode 4.6
* Template = Cocoa Touch Static Library
* Name = fuego-on-ios

File changes:

* Remove all files that were automatically added to the project by the Xcode template

NEWS
This file contains information about Fuego releases. It also contains a
(probably incomplete) list of the most important changes or of all
critical bugfixes in releases from a bugfix branch. For a detailed changelog
of changes in the main development branch, consult the version control
history.

Current development version (branch VERSION_1_FIXES)
====================================================

* update configure.ac for current autotools
* fixed to compile on clang-600.0.54
* fixed to compile without warnings on gcc 4.6.3 on Ubuntu

Version 1.1 - 2011 Mar 11
=========================

* Added missing doxygen documentation source files to distribution
* Fixed compilation error if HOST_CPU was undefined
* Set default for lock-free search to true if HOST_CPU is i686
* Implement cgos-gameover to fix unnecessary pondering with CGOS Python client
  after game is over
* Avoid a crash if more than 2 GB are allocated on 32-bit Windows systems
  without support for large memory addresses (tickets #73, #76)
  The default memory used is now limited to 1 GB and the number od threads
  to 4. The expand threshold is set to 3.
* Updated documentation for 64-bit compilation on Windows (ticket #73)
* configure.ac: removed GCC warning flags -Wconversion, -ansi, -pedantic,
  because they generate warnings with Boost 1.45
* Check for existence of sys/sysctl.h with autoconf and use it in SgPlatform
  only if it exists.

Version 1.0 - 2011 Jan 23
=========================

* Command line option parsing is now done with boost::program_options.
  Deprecated SgCmdLineOpt is no longer used. Options start with "--". 
* Game-independent autobook builder to generate opening books
* Support for propagating proven wins and losses in the UCT tree
* Initial support for static win/loss detection in Go
* New configure parameter --enable-max-size. The parameter --enable-optimize
  is no longer supported. The new script setup-build.sh can be used to set up
  some commonly used build configurations.
* Use only the same floating point type for all counts and values in
  SgUctSearch. The floating point type cane be set with the configure
  parameter --enable-uct-value-type
* Added generic argument parsing function GtpCommand::Arg<T> and declared
  old function (GtpCommand::IntArg() etc.) as deprecated
* Old class GoGame was removed, GoGameRecord renamed to GoGame. GoGame
  does no longer provide non-const access to the tree and game state to
  guarantee its class invariants
* Made GoPlayer to store its search traces instead of appending them directly
  to the game tree
* Several optimizations to speed up the Monte-Carlo searches.
* Fuego now can be compiled natively on Windows with Visual C++ and MinGW.
* The Monte-Carlo search automatically sets reasonable default values for
  memory, number of threads and lock-free mode depending on the current
  machine.
* New parameters "go_param use_book" and "uct_param_player max_ponder_time"
* Updated opening books
* Play forced opening moves on board size 13 as well
* Fixed bug in SgStatisticsExt::Variance()
* Support for multiple knowledge thresholds
* Renamed some classes (SgSList to SgArrayList, SgProvenNodeType to
  SgUctProvenType, SgMoveInfo to SgUctMoveInfo, SgVectorUtility to
  SgVectorUtil, GoGtpTimeSettings to GoTimeSettings)
* reuse_subtree is now enabled by default. The interpretation of max_games
  changed such that it includes the count of simulations from reused trees.
* Fixes and improvements related to lock-free search, tree pruning, virtual
  loss and multiple playouts in SgUctSearch.
* Fix bug #58: Static safety solver misses some unconditionally alive groups
* Other small fixes and improvements.

Version 0.4.1 - 2009 Dec 23
===========================
[Note that the numbering for this version is misleading. This is not a
bugfix release from the stable branch after version 0.4 (VERSION_0_4_FIXES),
but a new release with new and incompatible features from the main branch.]

* Replaced linked list SgList and related functions by SgVector - a wrapper
  around std:vector with the same interface as SgList. Changed implementations
  that added or removed from front of list for efficiency.
* Added hooks for supporting MPI distributed memory implementations
* Add small opening book for 5x5
* Improved 9x9 opening book
* Added link to optional big 19x19 opening books at Ben Lambrechts' site
* Fixed several bugs in static safety solver
* Updated to compile with gcc 4.4.0
* Implemented SgBalancer to achieve roughly equal number of Black and White 
  moves of a specific type in playouts
* Implemented randomized use of RAVE for SelectChild
* Support for proven nodes in SgUctSearch
* Support for kgs-time_settings and different KGS overtime modes
* Refactored SgSearch, reduce size of huge functions. 
* In SgSearch::SearchEngine, use a stack instead of a vector for efficiency
* New max_memory option to uct_param_search makes Fuego easier to configure
* Implement forced opening moves in GoUctPlayer
* Set bias term constant to 0 by default on all boardsizes

Version 0.4 - 2009 May 22
=========================

* Remove SgStreamFmtRestorer; use Boost IO State Savers instead 
* Remove SgUctPriorKnowledge; Derived thread state class will supply knowledge
  values through the GenerateAllMoves() function.
* Change move count type in SgUctNode from float to size_t
* Add support for knowledge in the core of the uct tree

Version 0.3 - 2008 Dec 17
=========================

* Change some default settings to be more friendly to end users (e.g.
  automatic loading of opening book, set ignore_clock to 0)
* New parameter GoUctPlayer::ResignMinGames
* Filter moves on empty edge in GoUctDefaultRootFilter
* Added fillboard and nakade heuristic as described in recent MoGo paper
  (not enabled by default)
* reg_genmove uses now the time limit set with go_param timelimit
* Use weighted RAVE updates in SgUctSearch
* Prune nodes with low visit counts if the tree runs out of memory

boost/boost.sh
#===============================================================================
# Filename:  boost.sh
# Author:    Pete Goodliffe
# Copyright: (c) Copyright 2009 Pete Goodliffe
# Licence:   Please feel free to use this, with attribution
#===============================================================================
#
# Builds a Boost framework for the iPhone.
# Creates a set of universal libraries that can be used on an iPhone and in the
# iPhone simulator. Then creates a pseudo-framework to make using boost in Xcode
# less painful.
#
# To configure the script, define:
#    BOOST_LIBS:        which libraries to build
#    IPHONEOS_BASESDK_VERSION: iPhone SDK version (e.g. 5.1); if left
#                       undefined, the latest SDK known to your Xcode will
#                       will be used
#    IPHONE_SIMULATOR_BASESDK_VERSION: iPhone Simulator SDK version (e.g.
#                       5.1); if left undefined, the latest SDK known to
#                       your Xcode will be used
#
# Then go get the source tar.bz of the boost you want to build, shove it in the
# same directory as this script, and run "./boost.sh". Grab a cuppa. And voila.
#===============================================================================

IPHONEOS_SDKPREFIX="iphoneos"
IPHONE_SIMULATOR_SDKPREFIX="iphonesimulator"

: ${BOOST_LIBS:="thread filesystem program_options system date_time"}
: ${IPHONEOS_BASESDK_VERSION:=`xcrun --sdk $IPHONEOS_SDKPREFIX --show-sdk-version`}
: ${IPHONEOS_DEPLOYMENT_TARGET:=7.0}
: ${IPHONE_SIMULATOR_BASESDK_VERSION:=`xcrun --sdk $IPHONE_SIMULATOR_SDKPREFIX --show-sdk-version`}
: ${IPHONE_SIMULATOR_DEPLOYMENT_TARGET:=7.0}
: ${XCODE_ROOT:=`xcode-select -print-path`}
: ${EXTRA_CPPFLAGS:="-DBOOST_AC_USE_PTHREADS -DBOOST_SP_USE_PTHREADS -std=gnu++98 -stdlib=libstdc++"}

# The EXTRA_CPPFLAGS definition works around a thread race issue in
# shared_ptr. I encountered this historically and have not verified that
# the fix is no longer required. Without using the posix thread primitives
# an invalid compare-and-swap ARM instruction (non-thread-safe) was used for the
# shared_ptr use count causing nasty and subtle bugs.
#
# Should perhaps also consider/use instead: -BOOST_SP_USE_PTHREADS

: ${SRCDIR:=`pwd`}
: ${IOSBUILDDIR:=`pwd`/ios/build}
: ${PREFIXDIR:=`pwd`/ios/prefix}
: ${IOSFRAMEWORKDIR:=`pwd`/ios/framework}
: ${COMPILER:="clang"}

BOOST_SRC=$SRCDIR/boost-trunk

#===============================================================================

# These variables are known to the Apple compiler. At least
# IPHONEOS_DEPLOYMENT_TARGET is, and if it remains exported it will be seen
# by the compiler and cause the bjam build to fail. To prevent this, we
# un-export the variables so that they stay hidden from the compiler.
export -n IPHONEOS_DEPLOYMENT_TARGET
export -n IPHONE_SIMULATOR_DEPLOYMENT_TARGET

IPHONEOS_SDKNAME="${IPHONEOS_SDKPREFIX}${IPHONEOS_BASESDK_VERSION}"
IPHONE_SIMULATOR_SDKNAME="${IPHONE_SIMULATOR_SDKPREFIX}${IPHONE_SIMULATOR_BASESDK_VERSION}"

PLATFORMS_BASEDIR="$XCODE_ROOT/Platforms"
IPHONEOS_PLATFORMDIR="$PLATFORMS_BASEDIR/iPhoneOS.platform"
IPHONE_SIMULATOR_PLATFORMDIR="$PLATFORMS_BASEDIR/iPhoneSimulator.platform"

IPHONEOS_BJAM_TOOLSET="${IPHONEOS_BASESDK_VERSION}~iphone"
IPHONE_SIMULATOR_BJAM_TOOLSET="${IPHONE_SIMULATOR_BASESDK_VERSION}~iphonesim"

IPHONEOS_CPPFLAGS="-miphoneos-version-min=$IPHONEOS_DEPLOYMENT_TARGET $EXTRA_CPPFLAGS"
IPHONE_SIMULATOR_CPPFLAGS="-mios-simulator-version-min=$IPHONE_SIMULATOR_DEPLOYMENT_TARGET $EXTRA_CPPFLAGS"

ARM_LIPO="$(xcrun -sdk $IPHONEOS_SDKNAME -find lipo)"
SIM_LIPO="$(xcrun -sdk $IPHONEOS_SDKNAME -find lipo)"
ARM_AR="$(xcrun -sdk $IPHONE_SIMULATOR_SDKNAME -find ar)"
SIM_AR="$(xcrun -sdk $IPHONE_SIMULATOR_SDKNAME -find ar)"
ARM_COMPILER="$(xcrun -sdk $IPHONE_SIMULATOR_SDKNAME -find $COMPILER)"
SIM_COMPILER="$(xcrun -sdk $IPHONE_SIMULATOR_SDKNAME -find $COMPILER)"

ARM_COMBINED_LIB=$IOSBUILDDIR/lib_boost_arm.a
SIM_COMBINED_LIB=$IOSBUILDDIR/lib_boost_x86.a

#===============================================================================


#===============================================================================
# Functions
#===============================================================================

abort()
{
    echo
    echo "Aborted: $@"
    exit 1
}

doneSection()
{
    echo
    echo "    ================================================================="
    echo "    Done"
    echo
}

#===============================================================================

cleanEverythingReadyToStart()
{
    echo Cleaning everything before we start to build...
	rm -rf iphone-build iphonesim-build
    rm -rf $IOSBUILDDIR
    rm -rf $PREFIXDIR
    rm -rf $IOSFRAMEWORKDIR/$FRAMEWORK_NAME.framework
    doneSection
}

#===============================================================================
updateBoost()
{
    echo Updating boost into $BOOST_SRC...

	# No need to git fetch, we expect that the submodule is up-to-date

	pushd $BOOST_SRC >/dev/null
	# Remove everything not under version control...
	git clean -dfx
	# Throw away local changes (i.e. modifications to user-config.jam)
	git reset --hard
	popd >/dev/null
	
	# These patches are required to make Boost 1.55.0 build with Clang 3.4 and
	# later (Xcode 5.1 uses Clang 3.4). The list of patches that is required
	# comes from this MacPorts ticket: https://trac.macports.org/ticket/42282
	patch --forward -p1 <01-atomic-patch-6bb71fd.diff
	patch --forward -p1 <02-atomic-patch-e4bde20.diff

	cat >> $BOOST_SRC/tools/build/v2/user-config.jam <<EOF
using darwin : $IPHONEOS_BJAM_TOOLSET
   : $ARM_COMPILER -arch armv7 -arch armv7s -arch arm64 -fvisibility=hidden -fvisibility-inlines-hidden $IPHONEOS_CPPFLAGS
   : <striper> <root>$IPHONEOS_PLATFORMDIR/Developer
   : <architecture>arm <target-os>iphone
   ;
using darwin : $IPHONE_SIMULATOR_BJAM_TOOLSET
   : $SIM_COMPILER -arch i386 -arch x86_64 -fvisibility=hidden -fvisibility-inlines-hidden $IPHONE_SIMULATOR_CPPFLAGS
   : <striper> <root>$IPHONE_SIMULATOR_PLATFORMDIR/Developer
   : <architecture>x86 <target-os>iphone
   ;
EOF

    doneSection
}

#===============================================================================

inventMissingHeaders()
{
    # These files are missing in the ARM iPhoneOS SDK, but they are in the simulator.
    # They are supported on the device, so we copy them from x86 SDK to a staging area
    # to use them on ARM, too.
    echo Invent missing headers
    cp $IPHONE_SIMULATOR_PLATFORMDIR/Developer/SDKs/iPhoneSimulator${IPHONE_SIMULATOR_BASESDK_VERSION}.sdk/usr/include/{crt_externs,bzlib}.h $BOOST_SRC
}

#===============================================================================

bootstrapBoost()
{
    cd $BOOST_SRC
    BOOST_LIBS_COMMA=$(echo $BOOST_LIBS | sed -e "s/ /,/g")
    echo "Bootstrapping (with libs $BOOST_LIBS_COMMA)"
    ./bootstrap.sh --with-libraries=$BOOST_LIBS_COMMA
    doneSection
}

#===============================================================================

buildBoostForiPhoneOS()
{
    cd $BOOST_SRC
    
	# Install this one so we can copy the includes for the frameworks...
    ./bjam -j16 --build-dir=../iphone-build --stagedir=../iphone-build/stage --prefix=$PREFIXDIR toolset=darwin-$IPHONEOS_BJAM_TOOLSET architecture=arm target-os=iphone macosx-version=iphone-${IPHONEOS_BASESDK_VERSION} define=_LITTLE_ENDIAN link=static stage
    ./bjam -j16 --build-dir=../iphone-build --stagedir=../iphone-build/stage --prefix=$PREFIXDIR toolset=darwin-$IPHONEOS_BJAM_TOOLSET architecture=arm target-os=iphone macosx-version=iphone-${IPHONEOS_BASESDK_VERSION} define=_LITTLE_ENDIAN link=static install
    doneSection

    ./bjam -j16 --build-dir=../iphonesim-build --stagedir=../iphonesim-build/stage --toolset=darwin-$IPHONE_SIMULATOR_BJAM_TOOLSET architecture=x86 target-os=iphone macosx-version=iphonesim-${IPHONE_SIMULATOR_BASESDK_VERSION} link=static stage
	doneSection
}

#===============================================================================

scrunchAllLibsTogetherInOneLibPerPlatform()
{
	cd $SRCDIR

    echo Splitting all existing fat binaries...
    for NAME in $BOOST_LIBS; do
        LIB_FILENAME="libboost_$NAME.a"
        OBJ_DIRNAME="obj_$NAME"

        echo Decomposing $LIB_FILENAME...

        # Must have separate obj folders for each library, because separate libraries may
        # contain object files with the same name
        mkdir -p $IOSBUILDDIR/armv7/$OBJ_DIRNAME
        mkdir -p $IOSBUILDDIR/armv7s/$OBJ_DIRNAME
        mkdir -p $IOSBUILDDIR/arm64/$OBJ_DIRNAME
        mkdir -p $IOSBUILDDIR/i386/$OBJ_DIRNAME
        mkdir -p $IOSBUILDDIR/x86_64/$OBJ_DIRNAME

        $ARM_LIPO "iphone-build/stage/lib/$LIB_FILENAME" -thin armv7 -o $IOSBUILDDIR/armv7/$LIB_FILENAME
        $ARM_LIPO "iphone-build/stage/lib/$LIB_FILENAME" -thin armv7s -o $IOSBUILDDIR/armv7s/$LIB_FILENAME
        $ARM_LIPO "iphone-build/stage/lib/$LIB_FILENAME" -thin arm64 -o $IOSBUILDDIR/arm64/$LIB_FILENAME

        $SIM_LIPO "iphonesim-build/stage/lib/$LIB_FILENAME" -thin i386 -o $IOSBUILDDIR/i386/$LIB_FILENAME
        $SIM_LIPO "iphonesim-build/stage/lib/$LIB_FILENAME" -thin x86_64 -o $IOSBUILDDIR/x86_64/$LIB_FILENAME

        (cd $IOSBUILDDIR/armv7/$OBJ_DIRNAME; ar -x ../$LIB_FILENAME );
        (cd $IOSBUILDDIR/armv7s/$OBJ_DIRNAME; ar -x ../$LIB_FILENAME );
        (cd $IOSBUILDDIR/arm64/$OBJ_DIRNAME; ar -x ../$LIB_FILENAME );
        (cd $IOSBUILDDIR/i386/$OBJ_DIRNAME; ar -x ../$LIB_FILENAME );
        (cd $IOSBUILDDIR/x86_64/$OBJ_DIRNAME; ar -x ../$LIB_FILENAME );
    done

    echo "Linking each architecture into an uberlib => libboost.a"
    rm -f $IOSBUILDDIR/*/libboost.a
    echo ...armv7
    (cd $IOSBUILDDIR/armv7; $ARM_AR crus libboost.a obj_*/*.o; )
    echo ...armv7s
    (cd $IOSBUILDDIR/armv7s; $ARM_AR crus libboost.a obj_*/*.o; )
    echo ...arm64
    (cd $IOSBUILDDIR/arm64; $ARM_AR crus libboost.a obj_*/*.o; )
    echo ...i386
    (cd $IOSBUILDDIR/i386;  $SIM_AR crus libboost.a obj_*/*.o; )
    echo ...x86_64
    (cd $IOSBUILDDIR/x86_64;  $SIM_AR crus libboost.a obj_*/*.o; )
}

#===============================================================================
buildFramework()
{
	: ${1:?}
	FRAMEWORKDIR=$1
	BUILDDIR=$2

	VERSION_TYPE=Alpha
	FRAMEWORK_NAME=boost
	FRAMEWORK_VERSION=A

	FRAMEWORK_CURRENT_VERSION=$BOOST_VERSION
	FRAMEWORK_COMPATIBILITY_VERSION=$BOOST_VERSION

    FRAMEWORK_BUNDLE=$FRAMEWORKDIR/$FRAMEWORK_NAME.framework
    echo "Framework: Building $FRAMEWORK_BUNDLE from $BUILDDIR..."

    rm -rf $FRAMEWORK_BUNDLE

    echo "Framework: Setting up directories..."
    mkdir -p $FRAMEWORK_BUNDLE
    mkdir -p $FRAMEWORK_BUNDLE/Versions
    mkdir -p $FRAMEWORK_BUNDLE/Versions/$FRAMEWORK_VERSION
    mkdir -p $FRAMEWORK_BUNDLE/Versions/$FRAMEWORK_VERSION/Resources
    mkdir -p $FRAMEWORK_BUNDLE/Versions/$FRAMEWORK_VERSION/Headers
    mkdir -p $FRAMEWORK_BUNDLE/Versions/$FRAMEWORK_VERSION/Documentation

    echo "Framework: Creating symlinks..."
    ln -s $FRAMEWORK_VERSION               $FRAMEWORK_BUNDLE/Versions/Current
    ln -s Versions/Current/Headers         $FRAMEWORK_BUNDLE/Headers
    ln -s Versions/Current/Resources       $FRAMEWORK_BUNDLE/Resources
    ln -s Versions/Current/Documentation   $FRAMEWORK_BUNDLE/Documentation
    ln -s Versions/Current/$FRAMEWORK_NAME $FRAMEWORK_BUNDLE/$FRAMEWORK_NAME

    FRAMEWORK_INSTALL_NAME=$FRAMEWORK_BUNDLE/Versions/$FRAMEWORK_VERSION/$FRAMEWORK_NAME

    echo "Lipoing library into $FRAMEWORK_INSTALL_NAME..."
    $ARM_LIPO -create $BUILDDIR/*/libboost.a -o "$FRAMEWORK_INSTALL_NAME" || abort "Lipo $1 failed"

    echo "Framework: Copying includes..."
    cp -r $PREFIXDIR/include/boost/*  $FRAMEWORK_BUNDLE/Headers/

    echo "Framework: Creating plist..."
    cat > $FRAMEWORK_BUNDLE/Resources/Info.plist <<EOF
<?xml version="1.0" encoding="UTF-8"?>
<!DOCTYPE plist PUBLIC "-//Apple//DTD PLIST 1.0//EN" "http://www.apple.com/DTDs/PropertyList-1.0.dtd">
<plist version="1.0">
<dict>
	<key>CFBundleDevelopmentRegion</key>
	<string>English</string>
	<key>CFBundleExecutable</key>
	<string>${FRAMEWORK_NAME}</string>
	<key>CFBundleIdentifier</key>
	<string>org.boost</string>
	<key>CFBundleInfoDictionaryVersion</key>
	<string>6.0</string>
	<key>CFBundlePackageType</key>
	<string>FMWK</string>
	<key>CFBundleSignature</key>
	<string>????</string>
	<key>CFBundleVersion</key>
	<string>${FRAMEWORK_CURRENT_VERSION}</string>
</dict>
</plist>
EOF
    doneSection
}

#===============================================================================
# Execution starts here
#===============================================================================

mkdir -p $IOSBUILDDIR

cleanEverythingReadyToStart
updateBoost

BOOST_VERSION=`cd $BOOST_SRC; git describe --tags | sed -e 's/^.*\/Boost_\([^\/]*\)/\1/'`
echo "BOOST_VERSION:     $BOOST_VERSION"
echo "BOOST_LIBS:        $BOOST_LIBS"
echo "BOOST_SRC:         $BOOST_SRC"
echo "IOSBUILDDIR:       $IOSBUILDDIR"
echo "PREFIXDIR:         $PREFIXDIR"
echo "IOSFRAMEWORKDIR:   $IOSFRAMEWORKDIR"
echo "IPHONEOS_BASESDK_VERSION: $IPHONEOS_BASESDK_VERSION"
echo "IPHONE_SIMULATOR_BASESDK_VERSION: $IPHONE_SIMULATOR_BASESDK_VERSION"
echo "XCODE_ROOT:        $XCODE_ROOT"
echo "COMPILER:          $COMPILER"
echo

inventMissingHeaders
bootstrapBoost
buildBoostForiPhoneOS
scrunchAllLibsTogetherInOneLibPerPlatform
buildFramework $IOSFRAMEWORKDIR $IOSBUILDDIR

echo "Completed successfully"
echo "The framework is located here: $IOSFRAMEWORKDIR/$FRAMEWORK_NAME.framework"
echo "Enjoy."

#===============================================================================
